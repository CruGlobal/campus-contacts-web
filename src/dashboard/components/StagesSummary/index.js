import React, { useState } from 'react';
import styled from '@emotion/styled';
import { withTheme } from 'emotion-theming';
import PropTypes from 'prop-types';
import moment from 'moment';
import { useQuery } from 'react-apollo-hooks';
import { useTranslation } from 'react-i18next';

import RangePicker from '../RangePicker';
import notSure from '../../assets/icons/stage-not-sure.svg';
import uninterested from '../../assets/icons/stage-uninterested.svg';
import curious from '../../assets/icons/stage-curious.svg';
import forgiven from '../../assets/icons/stage-forgiven.svg';
import growing from '../../assets/icons/stage-growing.svg';
import guiding from '../../assets/icons/stage-guiding.svg';
import spiritual from '../../assets/icons/stage-spiritual-conversations.svg';
import gospel from '../../assets/icons/stage-gospel-presentations.svg';
import holySpirit from '../../assets/icons/stage-holy-spirit-conversations.svg';
import personal from '../../assets/icons/stage-personal-decisions.svg';
import discipleship from '../../assets/icons/stage-discipleship-conversations.svg';

const Stages = styled.div`
    display: flex;
    flex-direction: row;
    justify-content: space-around;
`;

const Stage = styled.div`
    display: flex;
    flex-direction: column;
    align-items: center;
`;

const Icon = styled.div`
    height: 52px;
    width: 52px;
    background-size: contain;
    background-repeat: no-repeat;

    &.not-sure {
        background-image: url(${notSure});
    }

    &.uninterested {
        background-image: url(${uninterested});
    }

    &.curious {
        background-image: url(${curious});
    }

    &.forgiven {
        background-image: url(${forgiven});
    }

    &.growing {
        background-image: url(${growing});
    }

    &.guiding {
        background-image: url(${guiding});
    }

    &.discipleship-conversations {
        background-image: url(${discipleship});
    }

    &.gospel-presentations {
        background-image: url(${gospel});
    }

    &.holy-spirit-conversations {
        background-image: url(${holySpirit});
    }

    &.personal-decisions {
        background-image: url(${personal});
    }

    &.spiritual-conversations {
        background-image: url(${spiritual});
    }
`;

const Title = styled.div`
    font-weight: 600;
    font-size: 12px;
    line-height: 14px;
    color: ${({ theme }) => theme.colors.secondary};
    margin-top: 17px;
    text-transform: uppercase;
    overflow-wrap: break-word;
    width: 110px;
    text-align: center;
`;

const Value = styled.div`
    font-weight: 600;
    font-size: 32px;
    line-height: 38px;
    color: ${({ theme }) => theme.colors.highlightDarker};
    margin-top: 9px;
`;

const SummaryWrapper = styled.div`
    display: flex;
    flex-direction: column;
`;

const Footer = styled.div`
    margin-top: 28px;
`;

const StagesSummary = ({ query, variables, mapData }) => {
    const [dates, setDates] = useState({
        startDate: moment().subtract(7, 'days'),
        endDate: moment(),
    });
    const { t } = useTranslation('insights');
    const { data, loading } = useQuery(query, { variables });

    if (loading) {
        return <SummaryWrapper>{t('loading')}</SummaryWrapper>;
    }

    const onDatesChange = ({ startDate, endDate }) => {
        setDates({ startDate, endDate });
    };

    return (
        <SummaryWrapper>
            <Stages>
                {mapData(data).map(entry => (
                    <Stage key={entry.stage}>
                        <Icon className={entry.icon} />
                        <Title>{entry.stage}</Title>
                        <Value>{entry.count ? entry.count : '-'}</Value>
                    </Stage>
                ))}
            </Stages>
            <Footer>
                <RangePicker
                    onDatesChange={onDatesChange}
                    startDate={dates.startDate}
                    endDate={dates.endDate}
                />
            </Footer>
        </SummaryWrapper>
    );
};

export default withTheme(StagesSummary);

StagesSummary.propTypes = {
    query: PropTypes.object.isRequired,
    variables: PropTypes.object.isRequired,
    mapData: PropTypes.func.isRequired,
};
