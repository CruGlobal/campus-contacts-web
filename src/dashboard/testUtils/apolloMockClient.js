import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { SchemaLink } from 'apollo-link-schema';
import { addMockFunctionsToSchema } from 'graphql-tools';
import { buildClientSchema } from 'graphql/utilities';
import { createMockerFromIntrospection } from 'fraql/mock';
import gql from 'fraql';

import introspectionQuery from '../../../schema.json';

import { globalMocks } from './globalMocks';

export const createApolloMockClient = (mocks = {}) => {
    const schema = buildClientSchema(introspectionQuery);

    addMockFunctionsToSchema({
        schema,
        mocks: { ...globalMocks, ...mocks },
    });

    return new ApolloClient({
        cache: new InMemoryCache(),
        link: new SchemaLink({ schema }),
    });
};

export const mockFragment = (fragmentDocument, options) => {
    return createMockerFromIntrospection(introspectionQuery, {
        mocks: globalMocks,
    }).mockFragment(gql(fragmentDocument.loc.source.body), options);
};
