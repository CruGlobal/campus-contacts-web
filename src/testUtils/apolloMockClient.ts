import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { SchemaLink } from 'apollo-link-schema';
import { addMockFunctionsToSchema } from 'graphql-tools';
// @ts-ignore
import { buildClientSchema } from 'graphql/utilities';
// @ts-ignore
import { createMockerFromIntrospection } from 'fraql/mock';
// @ts-ignore
import gql from 'fraql';

import introspectionQuery from '../../schema.json';

import { globalMocks } from './globalMocks';

export const createApolloMockClient = (mocks = {}) => {
    const schema = buildClientSchema(introspectionQuery);

    addMockFunctionsToSchema({
        schema,
        mocks: { ...globalMocks, ...mocks },
    });

    return new ApolloClient({
        cache: new InMemoryCache(),
        link: new SchemaLink({ schema }),
    });
};

export const mockFragment = (fragmentDocument: any, options: any) => {
    return createMockerFromIntrospection(introspectionQuery, {
        mocks: globalMocks,
    }).mockFragment(gql(fragmentDocument.loc.source.body), options);
};
