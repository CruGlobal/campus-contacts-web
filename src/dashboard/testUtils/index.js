import React, { ReactElement } from 'react';
import { ApolloProvider } from 'react-apollo-hooks';
import { render } from '@testing-library/react';

import { createApolloMockClient } from './apolloMockClient';
import defaultTheme from '../defaultTheme';
import { ThemeProvider } from 'emotion-theming';
import AppContext from '../appContext';

export const renderWithContext = (component, { mocks } = {}, context = {}) => {
    const mockApolloClient = createApolloMockClient(mocks);

    // Warning: don't call any functions in here that return new instances on every call.
    // All the props need to stay the same otherwise renderer won't work.
    const wrapper = ({ children }) => (
        <ApolloProvider client={mockApolloClient}>
            <AppContext.Provider value={context}>
                <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>
            </AppContext.Provider>
        </ApolloProvider>
    );

    const renderResult = render(component, { wrapper });

    return {
        ...renderResult,
        snapshot: () => {
            expect(renderResult.container).toMatchSnapshot();
        },
    };
};
