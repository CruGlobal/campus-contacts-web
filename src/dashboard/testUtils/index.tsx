import React, { ReactNode } from 'react';
import { ApolloProvider } from 'react-apollo-hooks';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'emotion-theming';

import defaultTheme from '../defaultTheme';
import AppContext from '../appContext';

import { createApolloMockClient } from './apolloMockClient';

export const renderWithContext = (
    component: any,
    { mocks: mocks = {}, appContext = {} } = {},
) => {
    const mockApolloClient = createApolloMockClient(mocks);

    // Warning: don't call any functions in here that return new instances on every call.
    // All the props need to stay the same otherwise renderer won't work.
    const wrapper = ({ children }: { children: ReactNode }) => (
        <ApolloProvider client={mockApolloClient}>
            <AppContext.Provider value={appContext}>
                <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>
            </AppContext.Provider>
        </ApolloProvider>
    );

    // @ts-ignore
    const renderResult = render(component, { wrapper });
    return {
        ...renderResult,
        snapshot: () => {
            expect(renderResult.container).toMatchSnapshot();
        },
    };
};
