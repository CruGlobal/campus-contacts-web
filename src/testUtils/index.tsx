import React, { ReactNode, ReactElement } from 'react';
import { ApolloProvider } from 'react-apollo-hooks';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'emotion-theming';
import { IMocks } from 'graphql-tools';

import defaultTheme from '../defaultTheme';
import {
    AppContext,
    AppContextValue,
    appContextDefaultValue,
} from '../appContext';

import { createApolloMockClient } from './apolloMockClient';

interface RenderWithContextParams {
    mocks?: IMocks;
    appContext?: AppContextValue;
}

export const renderWithContext = (
    component: ReactElement,
    {
        mocks: mocks = {},
        appContext = appContextDefaultValue,
    }: RenderWithContextParams = {},
) => {
    const mockApolloClient = createApolloMockClient(mocks);

    // Warning: don't call any functions in here that return new instances on every call.
    // All the props need to stay the same otherwise renderer won't work.
    const wrapper = ({ children }: { children?: ReactNode }) => (
        <ApolloProvider client={mockApolloClient}>
            <AppContext.Provider value={appContext}>
                <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>
            </AppContext.Provider>
        </ApolloProvider>
    );

    const renderResult = render(component, { wrapper });
    return {
        ...renderResult,
        snapshot: () => {
            expect(renderResult.container).toMatchSnapshot();
        },
    };
};
