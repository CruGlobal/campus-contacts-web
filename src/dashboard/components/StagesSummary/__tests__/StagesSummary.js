import React from 'react';
import gql from 'graphql-tag';

import { renderWithContext } from '../../../testUtils';
import StagesSummary from '../';

const QUERY = gql`
    query communityReport {
        communityReport {
            communityStagesReport {
                pathwayStage
                personalStepsCompletedCount
            }
        }
    }
`;

describe('<StagesSummary />', () => {
    it('should render properly', async () => {
        renderWithContext(
            <StagesSummary
                query={QUERY}
                variables={{
                    organizationId: 1,
                }}
                mapData={data =>
                    data.communityReport.communityStagesReport.map(entry => ({
                        stage: entry.pathwayStage,
                        icon: entry.pathwayStage
                            .toLowerCase()
                            .replace(' ', '-'),
                        count: entry.otherStepsCompletedCount,
                    }))
                }
            />,
        ).snapshot();
    });

    // TODO: Add a test with mocking data when GraphQL schema contains above query
});
