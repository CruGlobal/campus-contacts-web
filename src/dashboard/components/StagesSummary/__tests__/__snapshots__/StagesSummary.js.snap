// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StagesSummary /> should render properly 1`] = `
<div>
  <div
    class="css-1ohf0ui"
  >
    Loading...
  </div>
</div>
`;
